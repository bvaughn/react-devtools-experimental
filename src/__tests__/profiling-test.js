// @flow

import typeof ReactTestRenderer from 'react-test-renderer';
import type Bridge from 'src/bridge';
import type Store from 'src/devtools/store';

describe('profiling', () => {
  let React;
  let ReactDOM;
  let Scheduler;
  let SchedulerTracing;
  let TestRenderer: ReactTestRenderer;
  let bridge: Bridge;
  let store: Store;
  let utils;

  beforeEach(() => {
    utils = require('./utils');
    utils.beforeEachProfiling();

    bridge = global.bridge;
    store = global.store;
    store.collapseNodesByDefault = false;

    React = require('react');
    ReactDOM = require('react-dom');
    Scheduler = require('scheduler');
    SchedulerTracing = require('scheduler/tracing');
    TestRenderer = utils.requireTestRenderer();
  });

  it('should throw if importing older/unsupported data', () => {
    const {
      prepareImportedProfilingData,
    } = require('src/devtools/views/Profiler/utils');
    expect(() =>
      prepareImportedProfilingData(
        JSON.stringify({
          version: 0,
        })
      )
    ).toThrow('Unsupported profiler export version "0"');
  });

  describe('ProfilingSummary', () => {
    it('should be collected for each commit', async done => {
      const Parent = ({ count }) => {
        Scheduler.advanceTime(10);
        const children = new Array(count)
          .fill(true)
          .map((_, index) => <Child key={index} duration={index} />);
        return (
          <React.Fragment>
            {children}
            <MemoizedChild duration={1} />
          </React.Fragment>
        );
      };
      const Child = ({ duration }) => {
        Scheduler.advanceTime(duration);
        return null;
      };
      const MemoizedChild = React.memo(Child);

      const container = document.createElement('div');

      utils.act(() => ReactDOM.render(<Parent count={2} />, container));
      utils.act(() => store.startProfiling());
      utils.act(() => ReactDOM.render(<Parent count={3} />, container));
      utils.act(() => ReactDOM.render(<Parent count={1} />, container));
      utils.act(() => ReactDOM.render(<Parent count={0} />, container));
      utils.act(() => store.stopProfiling());

      let profilingSummary = null;

      function Suspender({ previousPofilingSummary, rendererID, rootID }) {
        profilingSummary = store.profilingCache.ProfilingSummary.read({
          rendererID,
          rootID,
        });
        if (previousPofilingSummary != null) {
          expect(profilingSummary).toEqual(previousPofilingSummary);
        } else {
          expect(profilingSummary).toMatchSnapshot('ProfilingSummary');
        }
        return null;
      }

      const rendererID = utils.getRendererID();
      const rootID = store.roots[0];

      await utils.actAsync(() =>
        TestRenderer.create(
          <React.Suspense fallback={null}>
            <Suspender
              previousPofilingSummary={null}
              rendererID={rendererID}
              rootID={rootID}
            />
          </React.Suspense>
        )
      );

      expect(profilingSummary).not.toBeNull();

      utils.exportImportHelper(bridge, store, rendererID, rootID);

      await utils.actAsync(() =>
        TestRenderer.create(
          <React.Suspense fallback={null}>
            <Suspender
              previousPofilingSummary={profilingSummary}
              rendererID={rendererID}
              rootID={rootID}
            />
          </React.Suspense>
        )
      );

      done();
    });
  });

  describe('CommitDetails', () => {
    it('should be collected for each commit', async done => {
      const Parent = ({ count }) => {
        Scheduler.advanceTime(10);
        const children = new Array(count)
          .fill(true)
          .map((_, index) => <Child key={index} duration={index} />);
        return (
          <React.Fragment>
            {children}
            <MemoizedChild duration={1} />
          </React.Fragment>
        );
      };
      const Child = ({ duration }) => {
        Scheduler.advanceTime(duration);
        return null;
      };
      const MemoizedChild = React.memo(Child);

      const container = document.createElement('div');

      utils.act(() => store.startProfiling());
      utils.act(() => ReactDOM.render(<Parent count={2} />, container));
      utils.act(() => ReactDOM.render(<Parent count={3} />, container));
      utils.act(() => ReactDOM.render(<Parent count={1} />, container));
      utils.act(() => ReactDOM.render(<Parent count={0} />, container));
      utils.act(() => store.stopProfiling());

      const allCommitDetails = [];

      function Suspender({
        commitIndex,
        previousCommitDetails,
        rendererID,
        rootID,
      }) {
        const commitDetails = store.profilingCache.CommitDetails.read({
          commitIndex,
          rendererID,
          rootID,
        });
        if (previousCommitDetails != null) {
          expect(commitDetails).toEqual(previousCommitDetails);
        } else {
          allCommitDetails.push(commitDetails);
          expect(commitDetails).toMatchSnapshot(
            `CommitDetails commitIndex: ${commitIndex}`
          );
        }
        return null;
      }

      const rendererID = utils.getRendererID();
      const rootID = store.roots[0];

      for (let commitIndex = 0; commitIndex < 4; commitIndex++) {
        await utils.actAsync(() => {
          TestRenderer.create(
            <React.Suspense fallback={null}>
              <Suspender
                commitIndex={commitIndex}
                previousCommitDetails={null}
                rendererID={rendererID}
                rootID={rootID}
              />
            </React.Suspense>
          );
        });
      }

      expect(allCommitDetails).toHaveLength(4);

      utils.exportImportHelper(bridge, store, rendererID, rootID);

      for (let commitIndex = 0; commitIndex < 4; commitIndex++) {
        await utils.actAsync(() => {
          TestRenderer.create(
            <React.Suspense fallback={null}>
              <Suspender
                commitIndex={commitIndex}
                previousCommitDetails={allCommitDetails[commitIndex]}
                rendererID={rendererID}
                rootID={rootID}
              />
            </React.Suspense>
          );
        });
      }

      done();
    });

    it('should calculate a self duration based on actual children (not filtered children)', async done => {
      store.componentFilters = [utils.createDisplayNameFilter('^Parent$')];

      const Grandparent = () => {
        Scheduler.advanceTime(10);
        return (
          <React.Fragment>
            <Parent key="one" />
            <Parent key="two" />
          </React.Fragment>
        );
      };
      const Parent = () => {
        Scheduler.advanceTime(2);
        return <Child />;
      };
      const Child = () => {
        Scheduler.advanceTime(1);
        return null;
      };

      utils.act(() => store.startProfiling());
      utils.act(() =>
        ReactDOM.render(<Grandparent />, document.createElement('div'))
      );
      utils.act(() => store.stopProfiling());

      let commitDetails = null;

      function Suspender({ commitIndex, rendererID, rootID }) {
        commitDetails = store.profilingCache.CommitDetails.read({
          commitIndex,
          rendererID,
          rootID,
        });
        expect(commitDetails).toMatchSnapshot(
          `CommitDetails with filtered self durations`
        );
        return null;
      }

      const rendererID = utils.getRendererID();
      const rootID = store.roots[0];

      await utils.actAsync(() => {
        TestRenderer.create(
          <React.Suspense fallback={null}>
            <Suspender
              commitIndex={0}
              rendererID={rendererID}
              rootID={rootID}
            />
          </React.Suspense>
        );
      });

      expect(commitDetails).not.toBeNull();

      done();
    });

    it('should calculate self duration correctly for suspended views', async done => {
      let data;
      const getData = () => {
        if (data) {
          return data;
        } else {
          throw new Promise(resolve => {
            data = 'abc';
            resolve(data);
          });
        }
      };

      const Parent = () => {
        Scheduler.advanceTime(10);
        return (
          <React.Suspense fallback={<Fallback />}>
            <Async />
          </React.Suspense>
        );
      };
      const Fallback = () => {
        Scheduler.advanceTime(2);
        return 'Fallback...';
      };
      const Async = () => {
        Scheduler.advanceTime(3);
        const data = getData();
        return data;
      };

      utils.act(() => store.startProfiling());
      await utils.actAsync(() =>
        ReactDOM.render(<Parent />, document.createElement('div'))
      );
      utils.act(() => store.stopProfiling());

      const allCommitDetails = [];

      function Suspender({ commitIndex, rendererID, rootID }) {
        const commitDetails = store.profilingCache.CommitDetails.read({
          commitIndex,
          rendererID,
          rootID,
        });
        allCommitDetails.push(commitDetails);
        expect(commitDetails).toMatchSnapshot(
          `CommitDetails with filtered self durations`
        );
        return null;
      }

      const rendererID = utils.getRendererID();
      const rootID = store.roots[0];

      for (let commitIndex = 0; commitIndex < 2; commitIndex++) {
        await utils.actAsync(() => {
          TestRenderer.create(
            <React.Suspense fallback={null}>
              <Suspender
                commitIndex={commitIndex}
                rendererID={rendererID}
                rootID={rootID}
              />
            </React.Suspense>
          );
        });
      }

      expect(allCommitDetails).toHaveLength(2);

      done();
    });
  });

  describe('FiberCommits', () => {
    it('should be collected for each rendered fiber', async done => {
      const Parent = ({ count }) => {
        Scheduler.advanceTime(10);
        const children = new Array(count)
          .fill(true)
          .map((_, index) => <Child key={index} duration={index} />);
        return (
          <React.Fragment>
            {children}
            <MemoizedChild duration={1} />
          </React.Fragment>
        );
      };
      const Child = ({ duration }) => {
        Scheduler.advanceTime(duration);
        return null;
      };
      const MemoizedChild = React.memo(Child);

      const container = document.createElement('div');

      utils.act(() => store.startProfiling());
      utils.act(() => ReactDOM.render(<Parent count={1} />, container));
      utils.act(() => ReactDOM.render(<Parent count={2} />, container));
      utils.act(() => ReactDOM.render(<Parent count={3} />, container));
      utils.act(() => store.stopProfiling());

      const allFiberCommits = [];

      function Suspender({
        fiberID,
        previousFiberCommits,
        rendererID,
        rootID,
      }) {
        const fiberCommits = store.profilingCache.FiberCommits.read({
          fiberID,
          rendererID,
          rootID,
        });
        if (previousFiberCommits != null) {
          expect(fiberCommits).toEqual(previousFiberCommits);
        } else {
          allFiberCommits.push(fiberCommits);
          expect(fiberCommits).toMatchSnapshot(
            `FiberCommits: element ${fiberID}`
          );
        }
        return null;
      }

      const rendererID = utils.getRendererID();
      const rootID = store.roots[0];

      for (let index = 0; index < store.numElements; index++) {
        await utils.actAsync(() => {
          const fiberID = store.getElementIDAtIndex(index);
          if (fiberID == null) {
            throw Error(`Unexpected null ID for element at index ${index}`);
          }
          TestRenderer.create(
            <React.Suspense fallback={null}>
              <Suspender
                fiberID={fiberID}
                previousFiberCommits={null}
                rendererID={rendererID}
                rootID={rootID}
              />
            </React.Suspense>
          );
        });
      }

      expect(allFiberCommits).toHaveLength(store.numElements);

      utils.exportImportHelper(bridge, store, rendererID, rootID);

      for (let index = 0; index < store.numElements; index++) {
        await utils.actAsync(() => {
          const fiberID = store.getElementIDAtIndex(index);
          if (fiberID == null) {
            throw Error(`Unexpected null ID for element at index ${index}`);
          }
          TestRenderer.create(
            <React.Suspense fallback={null}>
              <Suspender
                fiberID={fiberID}
                previousFiberCommits={allFiberCommits[index]}
                rendererID={rendererID}
                rootID={rootID}
              />
            </React.Suspense>
          );
        });
      }

      done();
    });
  });

  describe('Interactions', () => {
    it('should be collected for every traced interaction', async done => {
      const Parent = ({ count }) => {
        Scheduler.advanceTime(10);
        const children = new Array(count)
          .fill(true)
          .map((_, index) => <Child key={index} duration={index} />);
        return (
          <React.Fragment>
            {children}
            <MemoizedChild duration={1} />
          </React.Fragment>
        );
      };
      const Child = ({ duration }) => {
        Scheduler.advanceTime(duration);
        return null;
      };
      const MemoizedChild = React.memo(Child);

      const container = document.createElement('div');

      utils.act(() => store.startProfiling());
      utils.act(() =>
        SchedulerTracing.unstable_trace(
          'mount: one child',
          Scheduler.unstable_now(),
          () => ReactDOM.render(<Parent count={1} />, container)
        )
      );
      utils.act(() =>
        SchedulerTracing.unstable_trace(
          'update: two children',
          Scheduler.unstable_now(),
          () => ReactDOM.render(<Parent count={2} />, container)
        )
      );
      utils.act(() => store.stopProfiling());

      let interactions = null;

      function Suspender({ previousInteractions, rendererID, rootID }) {
        interactions = store.profilingCache.Interactions.read({
          rendererID,
          rootID,
        });
        if (previousInteractions != null) {
          expect(interactions).toEqual(previousInteractions);
        } else {
          expect(interactions).toMatchSnapshot('Interactions');
        }
        return null;
      }

      const rendererID = utils.getRendererID();
      const rootID = store.roots[0];

      await utils.actAsync(() =>
        TestRenderer.create(
          <React.Suspense fallback={null}>
            <Suspender
              previousInteractions={null}
              rendererID={rendererID}
              rootID={rootID}
            />
          </React.Suspense>
        )
      );

      expect(interactions).not.toBeNull();

      utils.exportImportHelper(bridge, store, rendererID, rootID);

      await utils.actAsync(() =>
        TestRenderer.create(
          <React.Suspense fallback={null}>
            <Suspender
              previousInteractions={interactions}
              rendererID={rendererID}
              rootID={rootID}
            />
          </React.Suspense>
        )
      );

      done();
    });
  });

  it('should remove profiling data when roots are unmounted', async () => {
    const Parent = ({ count }) =>
      new Array(count)
        .fill(true)
        .map((_, index) => <Child key={index} duration={index} />);
    const Child = () => <div>Hi!</div>;

    const containerA = document.createElement('div');
    const containerB = document.createElement('div');

    utils.act(() => {
      ReactDOM.render(<Parent key="A" count={3} />, containerA);
      ReactDOM.render(<Parent key="B" count={2} />, containerB);
    });

    utils.act(() => store.startProfiling());

    utils.act(() => {
      ReactDOM.render(<Parent key="A" count={4} />, containerA);
      ReactDOM.render(<Parent key="B" count={1} />, containerB);
    });

    utils.act(() => ReactDOM.unmountComponentAtNode(containerB));

    utils.act(() => ReactDOM.unmountComponentAtNode(containerA));

    utils.act(() => store.stopProfiling());

    // Assert all maps are empty
    store.assertExpectedRootMapSizes();
  });
});
