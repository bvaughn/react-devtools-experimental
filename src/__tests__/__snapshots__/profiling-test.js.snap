// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`profiling CommitDetails should be collected for each commit: CommitDetails commitIndex: 0 1`] = `
Object {
  "actualDurations": Map {
    1 => 12,
    2 => 12,
    3 => 0,
    4 => 1,
    5 => 1,
  },
  "commitIndex": 0,
  "interactions": Array [],
  "rootID": 1,
  "selfDurations": Map {
    1 => 0,
    2 => 10,
    3 => 0,
    4 => 1,
    5 => 1,
  },
}
`;

exports[`profiling CommitDetails should be collected for each commit: CommitDetails commitIndex: 1 1`] = `
Object {
  "actualDurations": Map {
    3 => 0,
    4 => 1,
    6 => 2,
    2 => 13,
    1 => 13,
  },
  "commitIndex": 1,
  "interactions": Array [],
  "rootID": 1,
  "selfDurations": Map {
    3 => 0,
    4 => 1,
    6 => 2,
    2 => 10,
    1 => 0,
  },
}
`;

exports[`profiling CommitDetails should be collected for each commit: CommitDetails commitIndex: 2 1`] = `
Object {
  "actualDurations": Map {
    3 => 0,
    2 => 10,
    1 => 10,
  },
  "commitIndex": 2,
  "interactions": Array [],
  "rootID": 1,
  "selfDurations": Map {
    3 => 0,
    2 => 10,
    1 => 0,
  },
}
`;

exports[`profiling CommitDetails should be collected for each commit: CommitDetails commitIndex: 3 1`] = `
Object {
  "actualDurations": Map {
    2 => 10,
    1 => 10,
  },
  "commitIndex": 3,
  "interactions": Array [],
  "rootID": 1,
  "selfDurations": Map {
    2 => 10,
    1 => 0,
  },
}
`;

exports[`profiling CommitDetails should be collected for each commit: imported data 1`] = `
Object {
  "commitDetails": Array [
    Object {
      "actualDurations": Map {
        1 => 12,
        2 => 12,
        3 => 0,
        4 => 1,
        5 => 1,
      },
      "commitIndex": 0,
      "interactions": Array [],
      "rootID": 1,
      "selfDurations": Map {
        1 => 0,
        2 => 10,
        3 => 0,
        4 => 1,
        5 => 1,
      },
    },
    Object {
      "actualDurations": Map {
        3 => 0,
        4 => 1,
        6 => 2,
        2 => 13,
        1 => 13,
      },
      "commitIndex": 1,
      "interactions": Array [],
      "rootID": 1,
      "selfDurations": Map {
        3 => 0,
        4 => 1,
        6 => 2,
        2 => 10,
        1 => 0,
      },
    },
    Object {
      "actualDurations": Map {
        3 => 0,
        2 => 10,
        1 => 10,
      },
      "commitIndex": 2,
      "interactions": Array [],
      "rootID": 1,
      "selfDurations": Map {
        3 => 0,
        2 => 10,
        1 => 0,
      },
    },
    Object {
      "actualDurations": Map {
        2 => 10,
        1 => 10,
      },
      "commitIndex": 3,
      "interactions": Array [],
      "rootID": 1,
      "selfDurations": Map {
        2 => 10,
        1 => 0,
      },
    },
  ],
  "interactions": Object {
    "interactions": Array [],
    "rootID": 1,
  },
  "profilingOperations": Map {
    1 => Array [
      Uint32Array [
        1,
        1,
        17,
        6,
        80,
        97,
        114,
        101,
        110,
        116,
        5,
        67,
        104,
        105,
        108,
        100,
        1,
        48,
        1,
        49,
        1,
        1,
        11,
        1,
        1,
        4,
        1,
        12000,
        1,
        2,
        5,
        1,
        0,
        1,
        0,
        4,
        2,
        12000,
        1,
        3,
        5,
        2,
        2,
        2,
        3,
        4,
        3,
        0,
        1,
        4,
        5,
        2,
        2,
        2,
        4,
        4,
        4,
        1000,
        1,
        5,
        8,
        2,
        2,
        2,
        0,
        4,
        5,
        1000,
      ],
      Uint32Array [
        1,
        1,
        8,
        5,
        67,
        104,
        105,
        108,
        100,
        1,
        50,
        1,
        6,
        5,
        2,
        2,
        1,
        2,
        4,
        6,
        2000,
        4,
        2,
        14000,
        3,
        2,
        4,
        3,
        4,
        6,
        5,
        4,
        1,
        14000,
      ],
      Uint32Array [
        1,
        1,
        0,
        2,
        2,
        6,
        4,
        4,
        2,
        11000,
        3,
        2,
        2,
        3,
        5,
        4,
        1,
        11000,
      ],
      Uint32Array [
        1,
        1,
        0,
        2,
        1,
        3,
      ],
    ],
  },
  "profilingSnapshots": Map {
    1 => Map {},
  },
  "profilingSummary": Object {
    "commitDurations": Array [
      12,
      13,
      10,
      10,
    ],
    "commitTimes": Array [
      12,
      25,
      35,
      45,
    ],
    "initialTreeBaseDurations": Map {},
    "interactionCount": 0,
    "rootID": 1,
  },
  "version": 3,
}
`;

exports[`profiling CommitDetails should calculate a self duration based on actual children (not filtered children): CommitDetails with filtered self durations 1`] = `
Object {
  "actualDurations": Map {
    1 => 16,
    2 => 16,
    3 => 1,
    5 => 1,
  },
  "commitIndex": 0,
  "interactions": Array [],
  "rootID": 1,
  "selfDurations": Map {
    1 => 0,
    2 => 10,
    3 => 1,
    5 => 1,
  },
}
`;

exports[`profiling CommitDetails should calculate self duration correctly for suspended views: CommitDetails with filtered self durations 1`] = `
Object {
  "actualDurations": Map {
    1 => 15,
    2 => 15,
    3 => 5,
    4 => 2,
  },
  "commitIndex": 0,
  "interactions": Array [],
  "rootID": 1,
  "selfDurations": Map {
    1 => 0,
    2 => 10,
    3 => 3,
    4 => 2,
  },
}
`;

exports[`profiling CommitDetails should calculate self duration correctly for suspended views: CommitDetails with filtered self durations 2`] = `
Object {
  "actualDurations": Map {
    5 => 3,
    3 => 3,
  },
  "commitIndex": 1,
  "interactions": Array [],
  "rootID": 1,
  "selfDurations": Map {
    5 => 3,
    3 => 0,
  },
}
`;

exports[`profiling FiberCommits should be collected for each rendered fiber: FiberCommits: element 2 1`] = `
Object {
  "commitDurations": Array [
    0,
    10,
    1,
    10,
    2,
    10,
  ],
  "fiberID": 2,
  "rootID": 1,
}
`;

exports[`profiling FiberCommits should be collected for each rendered fiber: FiberCommits: element 3 1`] = `
Object {
  "commitDurations": Array [
    0,
    0,
    1,
    0,
    2,
    0,
  ],
  "fiberID": 3,
  "rootID": 1,
}
`;

exports[`profiling FiberCommits should be collected for each rendered fiber: FiberCommits: element 4 1`] = `
Object {
  "commitDurations": Array [
    0,
    1,
  ],
  "fiberID": 4,
  "rootID": 1,
}
`;

exports[`profiling FiberCommits should be collected for each rendered fiber: FiberCommits: element 5 1`] = `
Object {
  "commitDurations": Array [
    1,
    1,
    2,
    1,
  ],
  "fiberID": 5,
  "rootID": 1,
}
`;

exports[`profiling FiberCommits should be collected for each rendered fiber: FiberCommits: element 6 1`] = `
Object {
  "commitDurations": Array [
    2,
    2,
  ],
  "fiberID": 6,
  "rootID": 1,
}
`;

exports[`profiling FiberCommits should be collected for each rendered fiber: imported data 1`] = `
Object {
  "commitDetails": Array [
    Object {
      "actualDurations": Map {
        1 => 11,
        2 => 11,
        3 => 0,
        4 => 1,
      },
      "commitIndex": 0,
      "interactions": Array [],
      "rootID": 1,
      "selfDurations": Map {
        1 => 0,
        2 => 10,
        3 => 0,
        4 => 1,
      },
    },
    Object {
      "actualDurations": Map {
        3 => 0,
        5 => 1,
        2 => 11,
        1 => 11,
      },
      "commitIndex": 1,
      "interactions": Array [],
      "rootID": 1,
      "selfDurations": Map {
        3 => 0,
        5 => 1,
        2 => 10,
        1 => 0,
      },
    },
    Object {
      "actualDurations": Map {
        3 => 0,
        5 => 1,
        6 => 2,
        2 => 13,
        1 => 13,
      },
      "commitIndex": 2,
      "interactions": Array [],
      "rootID": 1,
      "selfDurations": Map {
        3 => 0,
        5 => 1,
        6 => 2,
        2 => 10,
        1 => 0,
      },
    },
  ],
  "interactions": Object {
    "interactions": Array [],
    "rootID": 1,
  },
  "profilingOperations": Map {
    1 => Array [
      Uint32Array [
        1,
        1,
        15,
        6,
        80,
        97,
        114,
        101,
        110,
        116,
        5,
        67,
        104,
        105,
        108,
        100,
        1,
        48,
        1,
        1,
        11,
        1,
        1,
        4,
        1,
        11000,
        1,
        2,
        5,
        1,
        0,
        1,
        0,
        4,
        2,
        11000,
        1,
        3,
        5,
        2,
        2,
        2,
        3,
        4,
        3,
        0,
        1,
        4,
        8,
        2,
        2,
        2,
        0,
        4,
        4,
        1000,
      ],
      Uint32Array [
        1,
        1,
        8,
        5,
        67,
        104,
        105,
        108,
        100,
        1,
        49,
        1,
        5,
        5,
        2,
        2,
        1,
        2,
        4,
        5,
        1000,
        4,
        2,
        12000,
        3,
        2,
        3,
        3,
        5,
        4,
        4,
        1,
        12000,
      ],
      Uint32Array [
        1,
        1,
        8,
        5,
        67,
        104,
        105,
        108,
        100,
        1,
        50,
        1,
        6,
        5,
        2,
        2,
        1,
        2,
        4,
        6,
        2000,
        4,
        2,
        14000,
        3,
        2,
        4,
        3,
        5,
        6,
        4,
        4,
        1,
        14000,
      ],
    ],
  },
  "profilingSnapshots": Map {
    1 => Map {},
  },
  "profilingSummary": Object {
    "commitDurations": Array [
      11,
      11,
      13,
    ],
    "commitTimes": Array [
      11,
      22,
      35,
    ],
    "initialTreeBaseDurations": Map {},
    "interactionCount": 0,
    "rootID": 1,
  },
  "version": 3,
}
`;

exports[`profiling Interactions should be collected for every traced interaction: Interactions 1`] = `
Object {
  "interactions": Array [
    Object {
      "__count": 1,
      "commits": Array [
        0,
      ],
      "id": 0,
      "name": "mount: one child",
      "timestamp": 0,
    },
    Object {
      "__count": 0,
      "commits": Array [
        1,
      ],
      "id": 1,
      "name": "update: two children",
      "timestamp": 11,
    },
  ],
  "rootID": 1,
}
`;

exports[`profiling Interactions should be collected for every traced interaction: imported data 1`] = `
Object {
  "commitDetails": Array [
    Object {
      "actualDurations": Map {
        1 => 11,
        2 => 11,
        3 => 0,
        4 => 1,
      },
      "commitIndex": 0,
      "interactions": Array [
        Object {
          "__count": 1,
          "id": 0,
          "name": "mount: one child",
          "timestamp": 0,
        },
      ],
      "rootID": 1,
      "selfDurations": Map {
        1 => 0,
        2 => 10,
        3 => 0,
        4 => 1,
      },
    },
    Object {
      "actualDurations": Map {
        3 => 0,
        5 => 1,
        2 => 11,
        1 => 11,
      },
      "commitIndex": 1,
      "interactions": Array [
        Object {
          "__count": 0,
          "id": 1,
          "name": "update: two children",
          "timestamp": 11,
        },
      ],
      "rootID": 1,
      "selfDurations": Map {
        3 => 0,
        5 => 1,
        2 => 10,
        1 => 0,
      },
    },
  ],
  "interactions": Object {
    "interactions": Array [
      Object {
        "__count": 1,
        "commits": Array [
          0,
        ],
        "id": 0,
        "name": "mount: one child",
        "timestamp": 0,
      },
      Object {
        "__count": 0,
        "commits": Array [
          1,
        ],
        "id": 1,
        "name": "update: two children",
        "timestamp": 11,
      },
    ],
    "rootID": 1,
  },
  "profilingOperations": Map {
    1 => Array [
      Uint32Array [
        1,
        1,
        15,
        6,
        80,
        97,
        114,
        101,
        110,
        116,
        5,
        67,
        104,
        105,
        108,
        100,
        1,
        48,
        1,
        1,
        11,
        1,
        1,
        4,
        1,
        11000,
        1,
        2,
        5,
        1,
        0,
        1,
        0,
        4,
        2,
        11000,
        1,
        3,
        5,
        2,
        2,
        2,
        3,
        4,
        3,
        0,
        1,
        4,
        8,
        2,
        2,
        2,
        0,
        4,
        4,
        1000,
      ],
      Uint32Array [
        1,
        1,
        8,
        5,
        67,
        104,
        105,
        108,
        100,
        1,
        49,
        1,
        5,
        5,
        2,
        2,
        1,
        2,
        4,
        5,
        1000,
        4,
        2,
        12000,
        3,
        2,
        3,
        3,
        5,
        4,
        4,
        1,
        12000,
      ],
    ],
  },
  "profilingSnapshots": Map {
    1 => Map {},
  },
  "profilingSummary": Object {
    "commitDurations": Array [
      11,
      11,
    ],
    "commitTimes": Array [
      11,
      22,
    ],
    "initialTreeBaseDurations": Map {},
    "interactionCount": 2,
    "rootID": 1,
  },
  "version": 3,
}
`;

exports[`profiling ProfilingSummary should be collected for each commit: ProfilingSummary 1`] = `
Object {
  "commitDurations": Array [
    13,
    10,
    10,
  ],
  "commitTimes": Array [
    13,
    23,
    33,
  ],
  "initialTreeBaseDurations": Map {
    1 => 12,
    2 => 12,
    3 => 0,
    4 => 1,
    5 => 1,
  },
  "interactionCount": 0,
  "rootID": 1,
}
`;

exports[`profiling ProfilingSummary should be collected for each commit: imported data 1`] = `
Object {
  "commitDetails": Array [
    Object {
      "actualDurations": Map {
        3 => 0,
        4 => 1,
        6 => 2,
        2 => 13,
        1 => 13,
      },
      "commitIndex": 0,
      "interactions": Array [],
      "rootID": 1,
      "selfDurations": Map {
        3 => 0,
        4 => 1,
        6 => 2,
        2 => 10,
        1 => 0,
      },
    },
    Object {
      "actualDurations": Map {
        3 => 0,
        2 => 10,
        1 => 10,
      },
      "commitIndex": 1,
      "interactions": Array [],
      "rootID": 1,
      "selfDurations": Map {
        3 => 0,
        2 => 10,
        1 => 0,
      },
    },
    Object {
      "actualDurations": Map {
        2 => 10,
        1 => 10,
      },
      "commitIndex": 2,
      "interactions": Array [],
      "rootID": 1,
      "selfDurations": Map {
        2 => 10,
        1 => 0,
      },
    },
  ],
  "interactions": Object {
    "interactions": Array [],
    "rootID": 1,
  },
  "profilingOperations": Map {
    1 => Array [
      Uint32Array [
        1,
        1,
        8,
        5,
        67,
        104,
        105,
        108,
        100,
        1,
        50,
        1,
        6,
        5,
        2,
        2,
        1,
        2,
        4,
        6,
        2000,
        4,
        2,
        14000,
        3,
        2,
        4,
        3,
        4,
        6,
        5,
        4,
        1,
        14000,
      ],
      Uint32Array [
        1,
        1,
        0,
        2,
        2,
        6,
        4,
        4,
        2,
        11000,
        3,
        2,
        2,
        3,
        5,
        4,
        1,
        11000,
      ],
      Uint32Array [
        1,
        1,
        0,
        2,
        1,
        3,
      ],
    ],
  },
  "profilingSnapshots": Map {
    1 => Map {
      1 => Object {
        "children": Array [
          2,
        ],
        "displayName": null,
        "id": 1,
        "key": null,
        "type": 11,
      },
      2 => Object {
        "children": Array [
          3,
          4,
          5,
        ],
        "displayName": "Parent",
        "id": 2,
        "key": null,
        "type": 5,
      },
      3 => Object {
        "children": Array [],
        "displayName": "Child",
        "id": 3,
        "key": "0",
        "type": 5,
      },
      4 => Object {
        "children": Array [],
        "displayName": "Child",
        "id": 4,
        "key": "1",
        "type": 5,
      },
      5 => Object {
        "children": Array [],
        "displayName": "Child",
        "id": 5,
        "key": null,
        "type": 8,
      },
    },
  },
  "profilingSummary": Object {
    "commitDurations": Array [
      13,
      10,
      10,
    ],
    "commitTimes": Array [
      13,
      23,
      33,
    ],
    "initialTreeBaseDurations": Map {
      1 => 12,
      2 => 12,
      3 => 0,
      4 => 1,
      5 => 1,
    },
    "interactionCount": 0,
    "rootID": 1,
  },
  "version": 3,
}
`;
